### PR TITLE
Refactored how we start calico with kubelet context to cancel

### PR DESCRIPTION
**Pull request by luthermonson**
#### Proposed Changes ####

We needed to rework how we were handling the startup for Calico node and felix. We also discovered a few additional issues related to some tweaks that needed to be corrected.

#### Types of Changes ####

Bugfix

#### Verification ####

Deploy a windows node with no error in the log files.

#### Linked Issues ####

* https://github.com/rancher/rke2/issues/1638
* https://github.com/rancher/rke2/issues/1470

### CODE REVIEW

File: pkg/pebinaryexecutor/pebinary.go [hunk 1]
```diff
@@ -190,6 +200,19 @@ func (p *PEBinaryConfig) KubeProxy(args []string) error {
...
+		if strings.Contains(arg, "source-vip") {
+			si = i
+		}
+	}
+
+	si, err := argIndex(args, "source-vip")
+	if err != nil {
+		return err
+	}
+	args[si] = fmt.Sprintf("--source-vip=%s", vip)
```
Comment (briandowns) [inline pkg/pebinaryexecutor/pebinary.go:214]: Nit:

```suggestion
	args[si] = "--source-vip="+vip
```

This is more direct and removes a significant amount of unnecessary processing from "under the hood".

File: pkg/pebinaryexecutor/pebinary.go [hunk 2]
```diff
@@ -254,6 +277,15 @@ func getArgs(argsMap map[string]string) []string {
 	return args
 }
 
+func argIndex(args []string, s string) (int, error) {
+	for i, arg := range args {
+		if strings.Contains(arg, s) {
+			return i, nil
+		}
+	}
+	return -1, errors.New("index not found")
```
Comment (briandowns) [inline pkg/pebinaryexecutor/pebinary.go:286]: Can we provide a little more information in this message?

File: pkg/pebinaryexecutor/pebinary.go [hunk 3]
```diff
@@ -254,6 +277,15 @@ func getArgs(argsMap map[string]string) []string {
 	return args
 }
 
+func argIndex(args []string, s string) (int, error) {
+	for i, arg := range args {
+		if strings.Contains(arg, s) {
+			return i, nil
+		}
+	}
+	return -1, errors.New("no matching argument found")
```
Comment (briandowns) [inline pkg/pebinaryexecutor/pebinary.go:286]: What significance does `-1` have here? Can you create a constant with a descriptive name for this value?

Comment (phillipsj) [reply pkg/pebinaryexecutor/pebinary.go:286]: What should I return as a default if the value isn't found? What is norm for that?

Comment (briandowns) [reply pkg/pebinaryexecutor/pebinary.go:286]: I would return `0`. Anything else implies meaning and that it'd be checked. In this case, I'd assume the `err` will be checked and in that case, the calling code won't be checking the int value.

File: pkg/pebinaryexecutor/pebinary.go [hunk 4]
```diff
@@ -190,6 +200,19 @@ func (p *PEBinaryConfig) KubeProxy(args []string) error {
 	}
 
 	args = append(getArgs(extraArgs), args...)
+	var si int
+	for i, arg := range args {
+		if strings.Contains(arg, "source-vip") {
+			si = i
+		}
+	}
```
Comment (brandond) [inline pkg/pebinaryexecutor/pebinary.go:208]: This appears to do the same thing as `argIndex`, should this have been removed?

File: pkg/pebinaryexecutor/pebinary.go [hunk 5]
```diff
@@ -254,6 +277,15 @@ func getArgs(argsMap map[string]string) []string {
 	return args
 }
 
+func argIndex(args []string, s string) (int, error) {
+	for i, arg := range args {
+		if strings.Contains(arg, s) {
+			return i, nil
+		}
+	}
```
Comment (brandond) [inline pkg/pebinaryexecutor/pebinary.go:285]: If you're going to make a function for this, it should be safe to use for any arg - not just ones that you know aren't substrings of other args or possible arg values.
```suggestion
	s = "--" + s
	for i, arg := range args {
		parts := strings.SplitN(arg, "=", 2)
		if strings.HasPrefix(parts[0], s) {
			return i, nil
		}
	}
```

**Review by brandond — CHANGES_REQUESTED** [review]

File: pkg/pebinaryexecutor/pebinary.go (hunk 4 above)
Comment (phillipsj) [reply pkg/pebinaryexecutor/pebinary.go]: Nice catch, fixed and retested.

File: pkg/pebinaryexecutor/pebinary.go (hunk 5 above)
Comment (phillipsj) [reply pkg/pebinaryexecutor/pebinary.go:281]: This adds functionality beyond what is required at the moment. I am happy to add an issue to track it, I just feel that it adds too much to the scope to address the current need and may cause potential issues if extra logic is added. This is only be used in the context of the Windows agent and the binary executor.

File: pkg/pebinaryexecutor/pebinary.go [hunk 6]
```diff
@@ -254,6 +273,15 @@ func getArgs(argsMap map[string]string) []string {
...
+	return indexNotFound, errors.New("no matching argument found")
```
Comment (brandond) [inline pkg/pebinaryexecutor/pebinary.go:282]: I don't think this needs a constant. The convention is to return `nil` or the type default when err is non-nil. Just do `return 0, errors.New(...)` and drop the constant.

File: pkg/pebinaryexecutor/pebinary.go (hunk 5 above)
Comment (brandond) [reply pkg/pebinaryexecutor/pebinary.go:281]: I saw you had an earlier version of this that didn't use a function. If you have a function, make it work right and be reusable. If it's not reusable, or is implemented such that it only works for your specific use case, don't make it a function.

**Review by brandond — CHANGES_REQUESTED** [review]

File: pkg/pebinaryexecutor/pebinary.go (hunk 6 above)
Comment (phillipsj) [reply pkg/pebinaryexecutor/pebinary.go:282]: @briandowns suggest using a constant. What should I do?

Comment (briandowns) [reply pkg/pebinaryexecutor/pebinary.go:282]: The const was for the `-1` since it would generally imply some meaning. If it's just 0, that's fine.

Comment (phillipsj) [reply pkg/pebinaryexecutor/pebinary.go]: @brandond Method removed.

File: pkg/pebinaryexecutor/pebinary.go (hunk 5 above)
Comment (phillipsj) [reply pkg/pebinaryexecutor/pebinary.go]: @brandond function removed.

**Review by briandowns — APPROVED** [review]

**Review by brandond — APPROVED** [review]

**Pull request merged**